### ISSUE TITLE
How create two isolated NLPs?

### ISSUE THREAD

**Issue by GuillermoPena** [axa-group/nlp.js#780]
Hi Jesus.
I'm trying create many NLPs with diferents NER/NLU configurations in the same code. 
I'm doing something like this:

```javascript
const { dockStart } = require('@nlpjs/basic')

const go = async () => {

  const config1 = {  use: [ 'Basic', 'LangEs' ], settings: { ner: { threshold: 0.8 } } }
  const dock1 = await dockStart(config1) // Config for nlp1
  const nlp1 = dock1.get('nlp')
  nlp1.addNerRuleOptionTexts('es', 'main1', 'e1', [ 'hola' ])

  const config2 = {  use: [ 'Basic', 'LangEs' ], settings: { ner: { threshold: 0.9 } } }
  const dock2 = await dockStart(config2) // Config for nlp2
  const nlp2 = dock2.get('nlp')
  nlp2.addNerRuleOptionTexts('es', 'main2', 'e2', [ 'adios' ])

  console.log(nlp1.ner.rules)
  console.log(nlp2.ner.rules)
  /* It always returns:
  {
    es: {
      main1: { name: 'main1', type: 'enum', rules: [Array] },
      main2: { name: 'main2', type: 'enum', rules: [Array] }
    }
  }
  */

 console.log(nlp1.ner.settings.threshold) // 0.8
 console.log(nlp2.ner.settings.threshold) // 0.8
}

go()
```

Two NLPs, but boths return all NER rules!
NER threshold of nlp2 and nlp1 is nlp1 threshold!

Is dock returning a singleton and all nlps are the same? 
How can I create differents NLPs?

**Versions:**

"@nlpjs/basic": "^4.17.5",
Node version: v14.15.0

**Comment by jesus-seijas-sp** [comment]
Hello, 
For isolations NLP.js has the concept of **container**, and **dock** is an structure of containers that contains one named **default**, but you can create multiple containers, each one with their independant plugins, and also taking profit of the plugins of the parents.

So you can get what you want that way:

```javascript
const { dockStart } = require('@nlpjs/basic')

const go = async () => {
  const config = {
    use: ['LangEs'],
    childs: {
      bot1: {
        settings: {
          nlp: {},
          ner: {
            threshold: 0.8
          }
        },
        use: ['Basic'],
      },
      bot2: {
        settings: {
          nlp: {},
          ner: {
            threshold: 0.9
          }
        },
        use: ['Basic'],
      }
    }
  }
  const dock = await dockStart(config);
  const nlp1 = dock.getContainer('bot1').get('nlp');
  nlp1.addNerRuleOptionTexts('es', 'main1', 'e1', ['hola'])

  const nlp2 = dock.getContainer('bot2').get('nlp')
  nlp2.addNerRuleOptionTexts('es', 'main2', 'e2', ['adios'])

  console.log(nlp1.ner.rules)
  console.log(nlp2.ner.rules)
  console.log(nlp1.ner.settings.threshold) // 0.8
  console.log(nlp2.ner.settings.threshold) // 0.8
}

go()
```

![image](https://user-images.githubusercontent.com/15154218/103656105-79960500-4f68-11eb-9963-dc9da3055838.png)

If you take a look into the settings, the dock is creating 3 containers:
- default: that will have the plugin for the LangEs, because this plugin will be shared for both childs.
- bot1: it will have his own implementation of the basic plugins, so that way it will have it's own NLP, NER and NLG instances
- bot2: similar to bot 1 in this case

To retrieve a container you can use:
```javascript
dock.getContainer('<name of the container>');
```
This is also explained a little bit in the quickstart to have different chatbots in different ports:
https://github.com/axa-group/nlp.js/blob/master/docs/v4/quickstart.md#different-port-for-microsoft-bot-framework-and-webchat

**Comment by GuillermoPena** [comment]
Perfect. Working! 
Thanks again Jesus!

**Comment by wickens** [comment]
Hi Jesús, does this mean I need to know the number of NLU models I need in advance (when the config is defined)? is there a way for me to create new NLU models dynamically at runtime?

My initial thinking was to wrap 

`const dock = await dockStart({ use: ["Basic"] });`

Inside my own manager class, and create a new instance of the manager class each time. I'm finding when I train, subsequent instances reach a lower number of epochs.

**Comment by jesus-seijas-sp** [comment]
Hello @wickens ,
dock instance is intended to create the structure to create bots, and with the configuration you can say how to create the containers for the chatbots and the different corpus per chatbot/language/domain.

If you don't know how many chatbots/languages/domains you have, well, then you can go to the low levels:
- Class Nlp from package @nlpjs/nlp, code located here: https://github.com/axa-group/nlp.js/blob/master/packages/nlp/src/nlp.js
   This class is able to create the language guesser, the slots, the ner, the nlg and the nlu. For the nlu it creates an structure by language (DomainManager), and by language you can have 1 or many domains. If you have only one domain, then only one model is trained. If you have more than one domain you can choose between training only one model or training domains+1 models, the +1 is because a master_domain is created than classify into domain instead of into intent, so when a sentence reach the DomainManager, first is classified to know what domain to use, and then the model of the domain returns the intent.
- Class NluManager from package @nlpjs/nlu, code located here: https://github.com/axa-group/nlp.js/blob/master/packages/nlu/src/nlu-manager.js
  This class is the NLU manager that class Nlp uses. Again, this class is multi-language and multi-domain, as explained before.
- Class DomainManager from package @nlpjs/nlu, code located here: https://github.com/axa-group/nlp.js/blob/master/packages/nlu/src/domain-manager.js
  This class represents a language, where a language can contain 1 or several domains, as explained before. That means that this one can have several models, but all these models will use the same language.
- Class NluNeural from package @nlpjs/nlu, code located here: https://github.com/axa-group/nlp.js/blob/master/packages/nlu/src/nlu-neural.js
  This class represents one model
- Class NeuralNetwork from package @nlpjs/neural, code located here: https://github.com/axa-group/nlp.js/blob/master/packages/neural/src/neural-network.js
  This class represents a Dense Neural Network, multi class (multi perceptron), but without language tools, so you have to provide the tokens (inputs) and labels (outputs/intents) in the given object format.

For each class you can take a look into the unit tests.

**Comment by wickens** [comment]
@jesus-seijas-sp  Thank you so much, that's really helped. I am now using the Nlp class and it seems to be working nicely. 

Not sure if this needs a new issue, but when I train I get the following error, however turning off auto save resolved it (I don't need to save the trained model). 

```
(node:10798) UnhandledPromiseRejectionWarning: Error: File cannot be written in web
    at /Users/marc/marc-node/nlp1/node_modules/@nlpjs/core/src/mock-fs.js:32:12
```
I am using the latest version, but see something similar was logged in #667

**Comment by jesus-seijas-sp** [comment]
@wickens This is because by default autoSave is true, so it's trying to save, but also it works in a browser and react native, so it does not have a file system registered by default. So 2 options:
- You want to train/run models in a browser or react native, or you don't want to autoSave models after training: 
```javascript
const nlp = new Nlp({ autoSave: false })
```
- You want to autoSave to filesystem:
```javascript
const { fs: requestfs } = require('@nlpjs/request');
const nlp = new Nlp();
nlp.container.register('fs', requestfs);
```

**Comment by wickens** [comment]
Makes sense now, thanks.
